### PR TITLE
Fix tests and make code simpler/clearer

### PR DESCRIPTION
**Pull request by bfred-it**
These commits avoid intermediary functions, shorten the code and deobfuscate the resolve/reject part of the code.

I'm not sure of how you feel about retouching code that works, but I thought I'd make a PR to possibly improve readability of the code.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Yeah, looks better :)

@jamestalmage LGTY?

File: index.js [hunk 1]
```diff
@@ -10,25 +10,23 @@ function generate(argIndex) {
...
 
-				setTimeout(function () {
-					complete(useValue ? value : result);
-				}, ms);
+		function thunk(result) {
+			return new Promise(function (resolve, reject) {
+				setTimeout(
+					willResolve ? resolve : reject,
+					ms,
+					useValue ? value : result
```
Comment (jamestalmage) [inline index.js]: The additional argument to `setTimeout` is not universal, and needs to be polyfilled on IE 9 and below. There is no data on mobil browser support. [Reference](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setTimeout)

File: index.js [hunk 2]
```diff
@@ -10,25 +10,23 @@ function generate(argIndex) {
...
 		thunk.catch = promise.catch.bind(promise);
-		return thunk;
 
-		function thunk(result) {
-			if (promise) {
-				// Prevent unhandled rejection errors if promise is never used, and only used as a thunk
-				promise.catch(function () {});
-			}
+		// Prevent unhandled rejection errors if promise is never used, and only used as a thunk
+		promise.catch(function () {});
```
Comment (jamestalmage) [inline index.js]: You can't do it this way.

We only want to suppress the `unhandledRejection` on the returned promise if it gets used as a `thunk`. The way you are doing it here, it suppresses it regardless.

We probably should write better tests

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: I'm not really concerned with IE support here. This module is mainly intended for Node.js.

Comment (jamestalmage) [inline index.js]: Except we use this all over our AVA tests, and when we add browser support their, we are going to start caring.

Comment (sindresorhus) [inline index.js]: Are we seriously ever going to support IE9, even with AVA?...

Comment (jamestalmage) [inline index.js]: Meh. I just checked, this works all the way back to Node 0.10. I thought it was a newer feature than that. I'm good with this.

File: index.js (hunk 2 above)
Comment (bfred-it) [inline index.js]: You are right about that but now I'm realizing that the previous code didn't work as intended in the first place.

From what I understand `unhandledrejection` will be dispatched pretty soon after the error happens and that `.catch` call won't happen until `thunk` is called again, which in this case will happen 5000ms after the promise was rejected (and `unhandledrejection` event dispatched)

```js
doSomethingFor5000ms()
.then(delay.reject(3000))
```

Comment (bfred-it) [inline index.js]: Test code:

```js
delay(5000)
  .then(delay.reject(10))
  .catch(function () {
    console.log('all good!')
  });
```

Old code output (`unhandledrejection` fired first, then the error icon changes to the checkmark after the 5000ms with `rejectionhandled`)

screen shot 2016-05-16 at 16 04 26

This PR's output:

screen shot 2016-05-16 at 16 06 45

Comment (bfred-it) [inline index.js]: In short, either we catch all of them, or catch none. This PR catches all of them, the stable code catches them only if the rejection happens not much after the `doSomethingFor5000ms()` ends, i.e.:

```js
// caught as expected
delay(5)
  .then(delay.reject(10))
  .catch(function () {
    console.log('all good!')
  });

// `unhandledrejection` and then `rejectionhandled`

delay(20)
  .then(delay.reject(10))
  .catch(function () {
    console.log('all good!')
  });
```

**Comment by jamestalmage** [comment]
That is the defined behavior of unhandledRejections. As long as `rejectionHandled` gets fired later, it's doing what it is supposed to.

**Comment by bfred-it** [comment]
Correct but then it's not really doing what the comment says: "Prevent unhandled rejection errors" since the error does happen

**Comment by jamestalmage** [comment]
Yes, the comment language is imprecise, but the behavior is correct.

See #12.

**Comment by bfred-it** [comment]
I'm not sure I can trust something that works half of the time. As a user of `.then(delay.reject(400))` I have no interest in the rejection of something I don't use, it's just noise in the console.

An alternative solution I think would be to make the internal promise lazy or semi-lazy. That is, don't resolve/reject anything internally until I either use `.then` or `.catch`.

**Comment by bfred-it** [comment]
How about this? The code might not be ideal for now but it works and there are never any exceptions to catch

**Comment by jamestalmage** [comment]
> I'm not sure I can trust something that works half of the time. 

Not sure what you mean by this. IMO, it works as designed. `unhandledRejections` are perfectly acceptable, as long as they are followed at some point in the future by a `rejectionHandled` event. The associated events was designed explicitly for this type of use.

>  I have no interest in the rejection of something I don't use, it's just noise in the console.

Yeah, that is unfortuneate. You will notice that chrome highlights it bright red when it rejects, then removes the highlighting, and changes the icon from an :x: to a ✔️ . Still, annoying if you are using it for production (I very rarely use it for that, mostly I use it for artificial delays in unit tests).

[`loud-rejection`](https://github.com/sindresorhus/loud-rejection) provides a much more sensible behavior for Node. `unhandledRejections` are only displayed when the program exits. You have until program exit to handle them. Obviously, the browser doesn't have a corollary for that. 

--

I don't think your latest commit changes anything, I think my previous comment still applies:

> We only want to suppress the `unhandledRejection` on the returned promise if it gets used as a thunk. The way you are doing it here, it suppresses it regardless.

-- 

@sindresorhus - What do you think? The only behavior this breaks, would be the one defined in [this unit test](https://github.com/jamestalmage/delay/blob/e96810406d775998ed0fbe87321dba8af7358d61/test.js#L90-L93). In other words, it breaks the `unhandledRejection` contract as followed:

```js
// If you never touch the promise - `unhandledRejection` will never fire. 
// While unlikely, this is exactly the sort of subtle bug `unhandledRejection` was intended to catch.
delay.reject(new Error('foo'));

// Simply attach a `.then` listener, and you get back the intended behavior.
// This will trigger `unhandledRejections`
delay.reject(new Error('foo')).then(() => {});

// Thunk usage as follows will eventually cause it to trigger:
Promise.resolve().then(delay.reject(new Error('foo')));

// With our current setup, the following will leave an unhandledRejection on the stack, 
// `bar` will remain unhandled, because the thunk is never called.
Promise.reject(new Error('foo')).then(delay.reject(new Error('bar'))).catch(() => {})
```

I think the last bit is troublesome. I could see people using it that way, and having confusing breakage.

-- 

I am starting to think the `promise`/`thunk` combo was too clever, and that we should have had separate functions for each.

**Comment by jamestalmage** [comment]
@bfred-it - Can you rebase? I've merged a test suite that verifies the current behavior. Your modifications won't pass some tests (so they will need to be modified), but they should fix the skipped test. At least this way we are making a conscious decision to modify the existing behavior.

Also, I think you should just revert 2aeaedb. I think it's really just a more verbose way of accomplishing the same thing you were before. If I'm missing something and that commit is doing something magical, please demonstrate with tests.

**Comment by bfred-it** [comment]
I changed the behavior of a test to make it consistent (i.e. avoid what's described in the _solution_ example here: https://github.com/sindresorhus/delay/pull/11#discussion_r63326151)

My thought yesterday was that maybe this problem would be avoided entirely by leaving `reject` out of `delay`. This feature can be replaced by:

```js
delay(400).then(e => Promise.reject(e));
```

So at this point I think we either take the feature out or avoid this combo.

**Comment by bfred-it** [comment]
Given that this wasn't going anywhere, I reduced the scope in #2 

My suggested solution was to use `delay(400).then(e => Promise.reject(e))` instead of `delay.reject(400)`, which I can already do.

**Pull request closed without merging**